replica/table.cc

```diff
@@ -685,103 +736,105 @@ future<> storage_group::split(compaction_group_list& list, sstables::compaction_
+                                          idx, schema()->ks_name(), schema()->cf_name()));
     }
 
-    return sg->split(_compaction_groups, split_compaction_options());
+    return sg->split(compaction_groups(), split_compaction_options());
+}
+
+future<> table::maybe_split_compaction_group_of(locator::tablet_id tablet_id) {
+    auto holder = async_gate().hold();
+    co_await _sg_manager->maybe_split_compaction_group_of(tablet_id.value());
 }
 
 std::unique_ptr<storage_group_manager> table::make_storage_group_manager() {
+    std::unique_ptr<storage_group_manager> ret;
     if (uses_tablets()) {
-        return std::make_unique<tablet_storage_group_manager>(*this, *_erm);
+        ret = std::make_unique<tablet_storage_group_manager>(*this, *_erm);
+    } else {
+        ret = std::make_unique<single_storage_group_manager>(*this);
     }
-    return std::make_unique<single_storage_group_manager>(*this);
+    ret->make_storage_groups();
+    return ret;
 }
 
 compaction_group* table::single_compaction_group_if_available() const noexcept {
-    return _compaction_groups.size() == 1 ? get_compaction_group(0) : nullptr;
+    return _sg_manager->single_compaction_group_if_available();
 }
 
 compaction_group* table::get_compaction_group(size_t id) const noexcept {
-    return _storage_groups[id]->main_compaction_group().get();
+    return storage_groups()[id]->main_compaction_group().get();
 }
 
 std::pair<size_t, locator::tablet_range_side>
 table::storage_group_of(dht::token token) const noexcept {
-    auto [idx, side] = _sg_manager->storage_group_of(token);
-    if (idx >= _storage_groups.size()) {
-        on_fatal_internal_error(tlogger, format("storage_group_for_token: index out of range: idx={} size_log2={} size={} token={}",
-                                                idx, _sg_manager->log2_storage_groups(), _storage_groups.size(), token));
-    }
-    auto& sg = *_storage_groups[idx];
-    if (!token.is_minimum() && !token.is_maximum() && !sg.token_range().contains(token, dht::token_comparator())) {
-        on_fatal_internal_error(tlogger, format("storage_group_for_token: storage_group idx={} range={} does not contain token={}",
-                 idx, sg.token_range(), token));
-    }
-    return {idx, side};
+    return _sg_manager->storage_group_of(token);
 }
 
 size_t table::storage_group_id_for_token(dht::token token) const noexcept {
-    return storage_group_of(token).first;
+    return _sg_manager->storage_group_id_for_token(token);
 }
 
 storage_group* table::storage_group_for_token(dht::token token) const noexcept {
-    return _storage_groups[storage_group_of(token).first].get();
+    return _sg_manager->storage_group_for_token(token);
 }
 
 compaction_group& table::compaction_group_for_token(dht::token token) const noexcept {
     auto [idx, range_side] = storage_group_of(token);
-    auto& sg = *_storage_groups[idx];
+    auto& sg = *storage_groups()[idx];
```

Comment:
Would be better not to expose storage_groups() at all, instead move the whole function to storage_group_manager. But not important for now.